File: src/DataCharts-System/frontend/src/services/apiClient.ts
/**
 * API客户端服务
 * 
 * 统一的API通信接口
 */

interface ApiResponse<T> {
  data: T
  status: number
  message?: string
}

interface UploadProgress {
  loaded: number
  total: number
  progress: number
}

class ApiClient {
  private baseURL: string
  private timeout: number

  constructor(baseURL: string = 'http://localhost:8000') {
    this.baseURL = baseURL
    this.timeout = 30000
  }

  private async request<T>(
    endpoint: string, 
    options: RequestInit = {}
  ): Promise<ApiResponse<T>> {
    const url = `${this.baseURL}${endpoint}`
    
    const defaultOptions: RequestInit = {
      headers: {
        'Content-Type': 'application/json',
      },
      ...options
    }

    try {
      const response = await fetch(url, defaultOptions)
      
      if (!response.ok) {
        throw new Error(`HTTP ${response.status}: ${response.statusText}`)
      }

      const data = await response.json()
      
      return {
        data,
        status: response.status,
        message: data.message || 'Success'
      }
    } catch (error) {
      console.error('API Request failed:', error)
      throw error
    }
  }

  // 数据相关API
  async uploadData(file: File, options?: any): Promise<ApiResponse<any>> {
    const formData = new FormData()
    formData.append('file', file)
    
    if (options) {
      formData.append('options', JSON.stringify(options))
    }

    return this.request('/api/data/upload', {
      method: 'POST',
      body: formData,
      headers: {} // 让浏览器自动设置Content-Type
    })
  }

  async uploadManualData(content: string, format: string): Promise<ApiResponse<any>> {
    return this.request('/api/data/manual', {
      method: 'POST',
      body: JSON.stringify({
        content,
        format,
        options: {}
      })
    })
  }

  async getData(dataId: string): Promise<ApiResponse<any>> {
    return this.request(`/api/data/${dataId}`)
  }

  async getDataPreview(dataId: string, rows: number = 10): Promise<ApiResponse<any>> {
    return this.request(`/api/data/${dataId}/preview?rows=${rows}`)
  }

  async processData(dataId: string, options: any): Promise<ApiResponse<any>> {
    return this.request('/api/data/process', {
      method: 'POST',
      body: JSON.stringify({
        data_id: dataId,
        preprocess_options: options
      })
    })
  }

  async listData(skip: number = 0, limit: number = 20): Promise<ApiResponse<any>> {
    return this.request(`/api/data/?skip=${skip}&limit=${limit}`)
  }

  async deleteData(dataId: string): Promise<ApiResponse<any>> {
    return this.request(`/api/data/${dataId}`, {
      method: 'DELETE'
    })
  }

  // 函数相关API
  async parseFunction(expression: string): Promise<ApiResponse<any>> {
    return this.request('/api/function/parse', {
      method: 'POST',
      body: JSON.stringify({ expression })
    })
  }

  async applyFunction(dataId: string, expression: string): Promise<ApiResponse<any>> {
    return this.request('/api/function/apply', {
      method: 'POST',
      body: JSON.stringify({
        data_id: dataId,
        expression
      })
    })
  }

  async getFunctionLibrary(): Promise<ApiResponse<any>> {
    return this.request('/api/function/library')
  }

  async validateFunction(expression: string): Promise<ApiResponse<any>> {
    return this.request('/api/function/validate', {
      method: 'POST',
      body: JSON.stringify({ expression })
    })
  }

  // 图表相关API
  async createChart(dataId: string, chartType: string, config: any): Promise<ApiResponse<any>> {
    return this.request('/api/chart/create', {
      method: 'POST',
      body: JSON.stringify({
        data_id: dataId,
        chart_type: chartType,
        config
      })
    })
  }

  async getChart(chartId: string): Promise<ApiResponse<any>> {
    return this.request(`/api/chart/${chartId}`)
  }

  async updateChart(chartId: string, dataId?: string, config?: any): Promise<ApiResponse<any>> {
    return this.request(`/api/chart/${chartId}`, {
      method: 'PUT',
      body: JSON.stringify({
        data_id: dataId,
        config
      })
    })
  }

  async exportChart(chartId: string, format: string = 'png'): Promise<Blob> {
    const response = await fetch(
      `${this.baseURL}/api/chart/${chartId}/export?format=${format}`
    )
    
    if (!response.ok) {
      throw new Error(`Export failed: ${response.statusText}`)
    }
    
    return response.blob()
  }

  async listCharts(skip: number = 0, limit: number = 20): Promise<ApiResponse<any>> {
    return this.request(`/api/chart/?skip=${skip}&limit=${limit}`)
  }

  async getChartTypes(): Promise<ApiResponse<any>> {
    return this.request('/api/chart/types')
  }

  async deleteChart(chartId: string): Promise<ApiResponse<any>> {
    return this.request(`/api/chart/${chartId}`, {
      method: 'DELETE'
    })
  }

  // 系统相关API
  async getSystemInfo(): Promise<ApiResponse<any>> {
    return this.request('/api/system/info')
  }

  async healthCheck(): Promise<ApiResponse<any>> {
    return this.request('/api/system/health')
  }

  async get版本(): Promise<ApiResponse<any>> {
    return this.request('/api/system/version')
  }

  // 工具方法
  async testConnection(): Promise<boolean> {
    try {
      await this.healthCheck()
      return true
    } catch {
      return false
    }
  }

  setBaseURL(url: string): void {
    this.baseURL = url
  }

  getBaseURL(): string {
    return this.baseURL
  }
}

// 创建默认实例
export const apiClient = new ApiClient()

// 导出类型和实例
export default ApiClient
export type { ApiResponse, UploadProgress }

File: src/DataCharts-System/desktop/src/utils/api_client.py
"""
API客户端模块

提供与后端服务通信的功能
"""

import requests
import json
from typing import Optional, Dict, Any, List
from PyQt6.QtCore import QObject, pyqtSignal, QThread


class APIClient(QObject):
    """API客户端类"""
    
    def __init__(self, base_url: str = "http://localhost:8000"):
        super().__init__()
        self.base_url = base_url.rstrip('/')
        self.session = requests.Session()
        self.session.headers.update({
            'Content-Type': 'application/json',
            'Accept': 'application/json'
        })
    
    def test_connection(self) -> Dict[str, Any]:
        """测试与后端的连接"""
        try:
            response = self.session.get(f"{self.base_url}/health", timeout=5)
            if response.status_code == 200:
                return {
                    "status": "success",
                    "message": "连接成功",
                    "data": response.json()
                }
            else:
                return {
                    "status": "error",
                    "message": f"服务器响应错误: {response.status_code}"
                }
        except requests.exceptions.ConnectionError:
            return {
                "status": "error",
                "message": "无法连接到后端服务，请确保服务已启动"
            }
        except requests.exceptions.Timeout:
            return {
                "status": "error",
                "message": "连接超时"
            }
        except Exception as e:
            return {
                "status": "error",
                "message": f"连接失败: {str(e)}"
            }
    
    def get_api_info(self) -> Dict[str, Any]:
        """获取API信息"""
        try:
            response = self.session.get(f"{self.base_url}/api/info", timeout=5)
            if response.status_code == 200:
                return {
                    "status": "success",
                    "data": response.json()
                }
            else:
                return {
                    "status": "error",
                    "message": f"获取API信息失败: {response.status_code}"
                }
        except Exception as e:
            return {
                "status": "error",
                "message": f"获取API信息失败: {str(e)}"
            }
    
    def upload_data(self, data: Any, file_info: Dict = None) -> Dict[str, Any]:
        """上传数据到后端"""
        try:
            payload = {
                "data": data,
                "file_info": file_info or {},
                "timestamp": str(self.get_current_timestamp())
            }
            
            response = self.session.post(
                f"{self.base_url}/api/data/upload",
                json=payload,
                timeout=30
            )
            
            if response.status_code == 200:
                return {
                    "status": "success",
                    "data": response.json()
                }
            else:
                return {
                    "status": "error",
                    "message": f"数据上传失败: {response.status_code}"
                }
        except Exception as e:
            return {
                "status": "error",
                "message": f"数据上传失败: {str(e)}"
            }
    
    def process_data(self, data: Any, processing_config: Dict) -> Dict[str, Any]:
        """处理数据"""
        try:
            payload = {
                "data": data,
                "config": processing_config
            }
            
            response = self.session.post(
                f"{self.base_url}/api/data/process",
                json=payload,
                timeout=60
            )
            
            if response.status_code == 200:
                return {
                    "status": "success",
                    "data": response.json()
                }
            else:
                return {
                    "status": "error",
                    "message": f"数据处理失败: {response.status_code}"
                }
        except Exception as e:
            return {
                "status": "error",
                "message": f"数据处理失败: {str(e)}"
            }
    
    def parse_function(self, expression: str) -> Dict[str, Any]:
        """解析函数表达式"""
        try:
            payload = {
                "expression": expression
            }
            
            response = self.session.post(
                f"{self.base_url}/api/function/parse",
                json=payload,
                timeout=10
            )
            
            if response.status_code == 200:
                return {
                    "status": "success",
                    "data": response.json()
                }
            else:
                return {
                    "status": "error",
                    "message": f"函数解析失败: {response.status_code}"
                }
        except Exception as e:
            return {
                "status": "error",
                "message": f"函数解析失败: {str(e)}"
            }
    
    def apply_function(self, expression: str, data: Any, variables: List = None) -> Dict[str, Any]:
        """应用函数到数据"""
        try:
            payload = {
                "expression": expression,
                "data": data,
                "variables": variables or []
            }
            
            response = self.session.post(
                f"{self.base_url}/api/function/apply",
                json=payload,
                timeout=60
            )
            
            if response.status_code == 200:
                return {
                    "status": "success",
                    "data": response.json()
                }
            else:
                return {
                    "status": "error",
                    "message": f"函数应用失败: {response.status_code}"
                }
        except Exception as e:
            return {
                "status": "error",
                "message": f"函数应用失败: {str(e)}"
            }
    
    def create_chart(self, chart_config: Dict) -> Dict[str, Any]:
        """创建图表"""
        try:
            response = self.session.post(
                f"{self.base_url}/api/chart/create",
                json=chart_config,
                timeout=30
            )
            
            if response.status_code == 200:
                return {
                    "status": "success",
                    "data": response.json()
                }
            else:
                return {
                    "status": "error",
                    "message": f"图表创建失败: {response.status_code}"
                }
        except Exception as e:
            return {
                "status": "error",
                "message": f"图表创建失败: {str(e)}"
            }
    
    def export_chart(self, chart_id: str, export_format: str = "png") -> Dict[str, Any]:
        """导出图表"""
        try:
            response = self.session.get(
                f"{self.base_url}/api/chart/{chart_id}/export",
                params={"format": export_format},
                timeout=30
            )
            
            if response.status_code == 200:
                return {
                    "status": "success",
                    "data": response.content if export_format in ["png", "jpg", "svg"] else response.json()
                }
            else:
                return {
                    "status": "error",
                    "message": f"图表导出失败: {response.status_code}"
                }
        except Exception as e:
            return {
                "status": "error",
                "message": f"图表导出失败: {str(e)}"
            }
    
    def get_current_timestamp(self):
        """获取当前时间戳"""
        from datetime import datetime
        return datetime.now().isoformat()


class APIWorker(QThread):
    """API调用工作线程"""
    
    finished = pyqtSignal(object)
    error = pyqtSignal(str)
    progress = pyqtSignal(int)
    
    def __init__(self, api_client: APIClient, method: str, *args, **kwargs):
        super().__init__()
        self.api_client = api_client
        self.method = method
        self.args = args
        self.kwargs = kwargs
    
    def run(self):
        """执行API调用"""
        try:
            self.progress.emit(10)
            
            # 获取API客户端方法
            api_method = getattr(self.api_client, self.method)
            if not api_method:
                self.error.emit(f"未找到API方法: {self.method}")
                return
            
            self.progress.emit(50)
            
            # 调用API方法
            result = api_method(*self.args, **self.kwargs)
            self.progress.emit(100)
            
            if result.get("status") == "success":
                self.finished.emit(result)
            else:
                self.error.emit(result.get("message", "API调用失败"))
                
        except Exception as e:
            self.error.emit(f"API调用异常: {str(e)}")


class APIManager(QObject):
    """API管理器"""
    
    connection_tested = pyqtSignal(object)
    
    def __init__(self, base_url: str = "http://localhost:8000"):
        super().__init__()
        self.api_client = APIClient(base_url)
        self.is_connected = False
    
    def test_connection_async(self):
        """异步测试连接"""
        self.worker = APIWorker(self.api_client, "test_connection")
        self.worker.finished.connect(self._on_connection_tested)
        self.worker.error.connect(self._on_connection_error)
        self.worker.start()
    
    def _on_connection_tested(self, result):
        """连接测试完成"""
        self.is_connected = result.get("status") == "success"
        self.connection_tested.emit(result)
    
    def _on_connection_error(self, error_msg):
        """连接测试失败"""
        self.is_connected = False
        self.connection_tested.emit({
            "status": "error",
            "message": error_msg
        })
    
    def get_client(self) -> APIClient:
        """获取API客户端"""
        return self.api_client
    
    def is_backend_available(self) -> bool:
        """检查后端是否可用"""
        return self.is_connected
